src/outputs/history.cpp

```diff
@@ -45,72 +45,116 @@ namespace parthenon {
-        PARTHENON_WARN(
-            "Resetting \"base\" MeshData to contain all blocks. This indicates that "
-            "the \"base\" MeshData container has been modified elsewhere. Double check "
-            "that the modification was intentional and is compatible with this reset.")
-        md_base->Set(pm->block_list, pm);
-      }
       auto result = hist_var.hst_fun(md_base.get());
-#ifdef MPI_PARALLEL
-      // need fence so that the result is ready prior to the MPI call
-      Kokkos::fence();
-      // apply separate chosen operations to each user-defined history output
-      MPI_Op usr_op;
-      switch (hist_var.hst_op) {
-      case UserHistoryOperation::sum:
-        usr_op = MPI_SUM;
-        break;
-      case UserHistoryOperation::max:
-        usr_op = MPI_MAX;
-        break;
-      case UserHistoryOperation::min:
-        usr_op = MPI_MIN;
-        break;
-      }
-      if (Globals::my_rank == 0) {
-        PARTHENON_MPI_CHECK(MPI_Reduce(MPI_IN_PLACE, &result, 1, MPI_PARTHENON_REAL,
-                                       usr_op, 0, MPI_COMM_WORLD));
-      } else {
-        PARTHENON_MPI_CHECK(MPI_Reduce(&result, &result, 1, MPI_PARTHENON_REAL, usr_op, 0,
-                                       MPI_COMM_WORLD));
+      results[hist_var.hst_op].push_back(result);
+      labels[hist_var.hst_op].push_back(hist_var.label);
+    }
+
+    // Check if the package has enrolled vector history functions which are stored in the
+    // Params under the `hist_vec_param_key` name.
+    if (!params.hasKey(hist_vec_param_key)) {
+      continue;
+    }
+    const auto &hist_vecs = params.Get<HstVec_list>(hist_vec_param_key);
+    for (const auto &hist_vec : hist_vecs) {
+      auto result = hist_vec.hst_vec_fun(md_base.get());
+      for (int n = 0; n < result.size(); n++) {
+        std::string label = hist_vec.label + std::to_string(n);
+        results[hist_vec.hst_op].push_back(result[n]);
+        labels[hist_vec.hst_op].push_back(label);
       }
-#endif
+    }
+  }
 
-      all_results.emplace_back(result);
-      all_labels.emplace_back(hist_var.label);
+  // Need fence so result is ready prior to MPI call
+  Kokkos::fence();
+
+#ifdef MPI_PARALLEL
+  for (auto &op : ops) {
+    if (results[op].size() == 0) {
+      continue;
+    }
+
+    MPI_Op usr_op;
+    if (op == UserHistoryOperation::sum) {
+      usr_op = MPI_SUM;
+    } else if (op == UserHistoryOperation::max) {
+      usr_op = MPI_MAX;
+    } else if (op == UserHistoryOperation::min) {
+      usr_op = MPI_MIN;
+    }
+
+    if (Globals::my_rank == 0) {
+      PARTHENON_MPI_CHECK(MPI_Reduce(MPI_IN_PLACE, &(results[op])[0], results[op].size(),
+                                     MPI_PARTHENON_REAL, usr_op, 0, MPI_COMM_WORLD));
+    } else {
+      PARTHENON_MPI_CHECK(MPI_Reduce(&(results[op])[0], &(results[op])[0],
+                                     results[op].size(), MPI_PARTHENON_REAL, usr_op, 0,
+                                     MPI_COMM_WORLD));
```

Comment:
I evaluate history results per-processor separately by MPI operation and then each MPI operation is called only once over that vector.

Reply:
this is really nice